File: BOOM/Boom.Model/Gameboard.cs
using System.Linq;
using TechnicalServices;

namespace Boom.Model
{
    public class GameBoard
    {
        public int Size { get; private set; }
        private SquareValues[,] Squares;
        private Ship[] Ships;
        private ILogger Logger;
        private IRandomGenerator RandomGenerator;

        public GameBoard(int size, Ship[] ships, ILogger logger, IRandomGenerator randomGenerator)
        {
            Size = size;
            Squares = new SquareValues[Size, Size];
            Logger = logger;
            RandomGenerator = randomGenerator;
            InitialiseEmptyBoard();
            Ships = ships;
        }

        //Sets all squares to empty
        private void InitialiseEmptyBoard()
        {
            for (int Row = 0; Row < Size; Row++)
            {
                for (int Column = 0; Column < Size; Column++)
                {
                    Squares[Row, Column] = SquareValues.Empty;
                }
            }
        }
        //Checks, in collaboration with Ships, whether any of them cover
        //the given row, column; if one does, invoke the Hit() method on it.
        public void CheckSquareAndRecordOutcome(int col, int row)
        {
            foreach (Ship ship in Ships)
            {
                if (ship.ShipOccupiesLocation(col, row))
                {
                    ship.Hit(col, row);
                    Squares[row, col] = SquareValues.Hit;
                    if (ship.IsSunk())
                    {
                        Logger.WriteLine(ship.Name + " sunk!");
                    }
                    else
                    {
                        Logger.WriteLine("Hit a " + ship.Name + " at (" + col + "," + row + ").");
                    }
                    if (CheckWin())
                    {
                        Logger.WriteLine("All ships sunk!");
                        Logger.WriteLine();
                    }
                    return;
                }
            }
            Squares[row, col] = SquareValues.Miss;
            Logger.WriteLine("Sorry, (" + col + "," + row + ") is a miss.");
        }

        //Returns true if the given position for the ship fits within the board 
        //and does not clash with another ship
        private bool IsValidPosition(Ship ship, int row, int col, Orientations orientation)
        {
            if (orientation == Orientations.Vertical && row + ship.Size > Size)
            {
                return false;
            }
            else if (orientation == Orientations.Horizontal && col + ship.Size > Size)
            {
                return false;
            }
            else
            {
                if (orientation == Orientations.Vertical)
                {
                    for (int Scan = 0; Scan < ship.Size; Scan++)
                    {
                        if (Squares[row + Scan, col] != SquareValues.Empty)
                        {
                            return false;
                        }
                    }
                }
                else if (orientation == Orientations.Horizontal)
                {
                    for (int Scan = 0; Scan < ship.Size; Scan++)
                    {
                        if (Squares[row, col + Scan] != SquareValues.Empty)
                        {
                            return false;
                        }
                    }
                }
            }
            return true;
        }
        //Returns true if all ships are sunk.
        public bool CheckWin()
        {
          return !Ships.Any(s => !s.IsSunk());
        }

        //Allows the actual array of squares to remain private
        public SquareValues ReadSquare(int row, int col)
        {
            return Squares[row, col];
        }

        //In collaboration with IsValidPosition, finds a random but valid
        //position for each of the ships set up, whether or not they already
        //have a position specified.
        public void RandomiseShipPlacement()
        {
            foreach (var ship in Ships)
            {
                Orientations orientation = 0; //default
                int row = 0;
                int col = 0;
                bool valid = false;
                while (valid == false)
                {
                    row = RandomGenerator.Next(0, Size);
                    col = RandomGenerator.Next(0, Size);
                    orientation = (Orientations)RandomGenerator.Next(0, 2);
                    valid = IsValidPosition(ship, row, col, orientation);
                }
                Logger.WriteLine("Computer placing the " + ship.Name);
                ship.SetPosition(row, col, orientation);
            }
        }
    }
}

File: BOOM.VB/Boom.Model/Gameboard.vb

Imports System.Linq
Imports TechnicalServices

Namespace Boom.Model
	Public Class GameBoard
		Public Property Size() As Integer
			Get
				Return m_Size
			End Get
			Private Set
				m_Size = Value
			End Set
		End Property
		Private m_Size As Integer
		Private Squares As SquareValues(,)
		Private Ships As Ship()
		Private Logger As ILogger
		Private RandomGenerator As IRandomGenerator

		Public Sub New(size__1 As Integer, ships__2 As Ship(), logger__3 As ILogger, randomGenerator__4 As IRandomGenerator)
			Size = size__1
			Squares = New SquareValues(Size - 1, Size - 1) {}
			Logger = logger__3
			RandomGenerator = randomGenerator__4
			InitialiseEmptyBoard()
			Ships = ships__2
		End Sub

		'Sets all squares to empty
		Private Sub InitialiseEmptyBoard()
			For Row As Integer = 0 To Size - 1
				For Column As Integer = 0 To Size - 1
					Squares(Row, Column) = SquareValues.Empty
				Next
			Next
		End Sub
		'Checks, in collaboration with Ships, whether any of them cover
		'the given row, column; if one does, invoke the Hit() method on it.
		Public Sub CheckSquareAndRecordOutcome(col As Integer, row As Integer)
			For Each ship As Ship In Ships
				If ship.ShipOccupiesLocation(col, row) Then
					ship.Hit(col, row)
					Squares(row, col) = SquareValues.Hit
					If ship.IsSunk() Then
						Logger.WriteLine(ship.Name + " sunk!")
					Else
                        Logger.WriteLine("Hit a " + ship.Name + " at (" + col.ToString() + "," + row.ToString() + ").")
                    End If
					If CheckWin() Then
						Logger.WriteLine("All ships sunk!")
						Logger.WriteLine()
					End If
					Return
				End If
			Next
			Squares(row, col) = SquareValues.Miss
            Logger.WriteLine("Sorry, (" + col.ToString() + "," + row.ToString() + ") is a miss.")
        End Sub

		'Returns true if the given position for the ship fits within the board 
		'and does not clash with another ship
		Private Function IsValidPosition(ship As Ship, row As Integer, col As Integer, orientation As Orientations) As Boolean
			If orientation = Orientations.Vertical AndAlso row + ship.Size > Size Then
				Return False
			ElseIf orientation = Orientations.Horizontal AndAlso col + ship.Size > Size Then
				Return False
			Else
				If orientation = Orientations.Vertical Then
					For Scan As Integer = 0 To ship.Size - 1
						If Squares(row + Scan, col) <> SquareValues.Empty Then
							Return False
						End If
					Next
				ElseIf orientation = Orientations.Horizontal Then
					For Scan As Integer = 0 To ship.Size - 1
						If Squares(row, col + Scan) <> SquareValues.Empty Then
							Return False
						End If
					Next
				End If
			End If
			Return True
		End Function
		'Returns true if all ships are sunk.
		Public Function CheckWin() As Boolean
			Return Not Ships.Any(Function(s) Not s.IsSunk())
		End Function

		'Allows the actual array of squares to remain private
		Public Function ReadSquare(row As Integer, col As Integer) As SquareValues
			Return Squares(row, col)
		End Function

		'In collaboration with IsValidPosition, finds a random but valid
		'position for each of the ships set up, whether or not they already
		'have a position specified.
		Public Sub RandomiseShipPlacement()
            For Each ship As Ship In Ships
                Dim orientation As Orientations = 0
                'default
                Dim row As Integer = 0
                Dim col As Integer = 0
                Dim valid As Boolean = False
                While valid = False
                    row = RandomGenerator.[Next](0, Size)
                    col = RandomGenerator.[Next](0, Size)
                    orientation = DirectCast(RandomGenerator.[Next](0, 2), Orientations)
                    valid = IsValidPosition(ship, row, col, orientation)
                End While
                Logger.WriteLine("Computer placing the " + ship.Name)
                ship.SetPosition(row, col, orientation)
            Next
        End Sub
	End Class
End Namespace

'=======================================================
'Service provided by Telerik (www.telerik.com)
'Conversion powered by NRefactory.
'Twitter: @telerik
'Facebook: facebook.com/telerik
'=======================================================
